src/shared/bpf-object.c

```diff
@@ -0,0 +1,139 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+
+#include <errno.h>
+#include <fcntl.h>
+#include <sys/resource.h>
+#include <unistd.h>
+
+#include "bpf-object.h"
+#include "bpf-program-v2.h"
+#include "fd-util.h"
+#include "set.h"
+
+int bpf_object_new(const unsigned char *mem_buf, size_t size, struct bpf_object **object) {
+        _cleanup_(bpf_object_freep) struct bpf_object *p = NULL;
+        /* Standard error code. */
+        long err;
+
+        assert(mem_buf);
+        assert(size > 0);
+        assert(object);
+
+        DECLARE_LIBBPF_OPTS(
+                        bpf_object_open_opts,
+                        opts);
+
+        p = bpf_object__open_mem(mem_buf, size, &opts);
+        /* Should either point to a valid BPF object or store standard error code. */
+        if (!p)
+                return -EINVAL;
+
+        err = libbpf_get_error(p);
```

Comment:
libbpf (at least currently) is using kernel's approach to encode errors as pointers for API that return pointers. libbpf_get_error() abstracts away underlying `PTR_ERR_OR_ZERO()` invocation. Not sure why it's declared long, but it's been like that since libbpf_get_error() got introduced few years ago. It's always either 0 or one of -Exxx error codes, so definitely fits into int. There should be no problem "truncating" this to int. All the existing code I'm aware of just uses `int err = libbpf_get_error(...);`, this won't break.